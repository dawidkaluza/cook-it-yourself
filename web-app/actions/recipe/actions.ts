"use server"

type Recipes = {
  items: [
    id: number,
    name: string,
    description: string
  ]
};

export const getMyRecipes = async () : Promise<Recipes> => {

};
