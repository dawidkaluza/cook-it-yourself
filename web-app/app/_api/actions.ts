"use server"

import {cookies} from "next/headers";

export function isSignedIn(){
  const cookieStore = cookies();
  const sessionCookie = process.env.NEXT_PUBLIC_API_GATEWAY_SESSION_COOKIE;
  return sessionCookie ? cookieStore.has(sessionCookie) : false;
}