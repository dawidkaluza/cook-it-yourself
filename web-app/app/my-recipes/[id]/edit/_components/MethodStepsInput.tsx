import React, {useState} from "react";
import {Step} from "@/app/my-recipes/_dtos/recipe";
import CryptoJS from 'crypto-js';

type Props = {
  errors?: string[];
  steps: Step[];
};

const MethodStepsInput = (props: Props) => {
  const [newStep, setNewStep] = useState("");
  const [steps, setSteps] = useState<Step[]>(props.steps);
  const [stepsToDelete, setStepsToDelete] = useState<Number[]>([]);

  const calcKey = (step: Step) => {
    return CryptoJS.MD5(step.text).toString();
  };

  const addStep = () => {
    if (newStep.trim().length === 0) {
      return;
    }

    const newSteps = [...steps];
    newSteps.push({ text: newStep });
    setSteps(newSteps);
    setNewStep("");
  };

  const deleteStep = (index: number) => {
    const step = steps[index];
    if (step.id) {
      const newStepsToDelete = [ ...stepsToDelete ];
      newStepsToDelete.push(step.id);
      setStepsToDelete(newStepsToDelete);
    }

    setSteps(
      steps.slice(0, index).concat(steps.slice(index + 1))
    );

    // TODO add shifting indexes down?
  };

  const errors = props.errors;
  return (
    <div className="row mb-4">
      <label htmlFor="newMethodStep" className="col-md-3 col-form-label text-md-end">Method steps</label>
      <div className="col-md-9">
        <div className="row">
          {stepsToDelete.map(stepId => (
            <input key={stepId.toString()}
                   type="hidden"
                   name="stepsToDelete"
                   value={stepId.toString()}
            />
          ))}
        </div>

        {steps.map((step, index) => (
          <div key={calcKey(step)} className="row">
            <MethodStepFields
              step={step}
              onDelete={() => deleteStep(index)}
            />
          </div>
        ))}

        <div className="row">
          <NewMethodStepFields value={newStep} onChange={setNewStep} onSubmit={addStep} />
        </div>

        <div className="row">
          {errors && errors.map(error => {
            return (
              <div key={error} className="invalid-feedback d-block">
                {error}
              </div>
            )
          })}
        </div>
      </div>
    </div>
  );
};

const MethodStepFields = (props: { step: Step, onDelete: () => void }) => {
  const { step, onDelete } = props;
  const { id } = step;
  const [ text, setText ] = useState(step.text);

  return (
    <div className="input-group">
      {id && (
        <input
          type="hidden"
          name="stepid"
          value={id}
        />
      )}

      <textarea
        name="methodStepText"
        className="form-control"
        placeholder="Method step"
        value={text}
        onChange={(event) => setText(event.target.value)}
        style={{height: "100px"}}
      />

      <div className="input-group-text">
        <button className="btn btn-sm btn-outline-danger" type="button" onClick={onDelete}>
          <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor"
               className="bi bi-dash-lg" viewBox="0 0 16 16">
            <path fillRule="evenodd" d="M2 8a.5.5 0 0 1 .5-.5h11a.5.5 0 0 1 0 1h-11A.5.5 0 0 1 2 8"/>
          </svg>
        </button>
      </div>
    </div>
  );
};

const NewMethodStepFields = (
  props: {
    value: string,
    onChange: (value: string) => void,
    onSubmit: () => void,
  }
) => {
  const {value, onChange, onSubmit} = props;
  return (
    <div className="input-group">
      <textarea
        name="newMethodStep" id="newMethodStep" className="form-control"
        value={value}
        placeholder="Instructions step-by-step to cook your recipe"
        style={{height: "100px"}}
        onChange={(e) => onChange(e.currentTarget.value)}
        onKeyDown={(e) => {
          if (e.key === 'Enter') {
            e.preventDefault();
            onSubmit();
          }
        }}
      />
      <div className="input-group-text">
        <button className="btn btn-sm btn-success" type="button" onClick={onSubmit}>
          <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor"
               className="bi bi-plus-lg" viewBox="0 0 16 16">
            <path fillRule="evenodd"
                  d="M8 2a.5.5 0 0 1 .5.5v5h5a.5.5 0 0 1 0 1h-5v5a.5.5 0 0 1-1 0v-5h-5a.5.5 0 0 1 0-1h5v-5A.5.5 0 0 1 8 2"/>
          </svg>
        </button>
      </div>
    </div>
  );
};

export {MethodStepsInput};