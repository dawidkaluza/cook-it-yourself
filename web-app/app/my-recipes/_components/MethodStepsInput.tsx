import React, {useState} from "react";
import {Step} from "@/app/my-recipes/_dtos/recipe";

type StepComponentData = Step & {
  key: number;
};

type MethodStepsInputProps = {
  errors?: string[];
  steps?: Step[];
};

const MethodStepsInput = (props: MethodStepsInputProps) => {
  const initialSteps = (props.steps ?? []).map((step, index) => {
    return {
      ...step,
      key: index + 1,
    };
  });

  const [newStep, setNewStep] = useState("");
  const [steps, setSteps] = useState<StepComponentData[]>(initialSteps);
  const [stepsToDelete, setStepsToDelete] = useState<number[]>([]);
  const [lastKey, setLastKey] = useState(initialSteps.length);

  // TODO maybe extract it to a separate hook?
  const generateKey = () => {
    const generatedKey = lastKey + 1;
    setLastKey(generatedKey);
    return generatedKey;
  };

  const addStep = () => {
    if (newStep.trim().length === 0) {
      return;
    }

    const newSteps = [...steps];
    newSteps.push({ text: newStep, key: generateKey() });
    setSteps(newSteps);
    setNewStep("");
  };

  const updateStep = (step: StepComponentData) => {
    const newSteps = [...steps];
    const index = newSteps.findIndex(filteredStep => filteredStep.key === step.key);
    newSteps[index] = step;
    setSteps(newSteps);
  };

  const deleteStep = (key: number) => {
    const index = steps.findIndex(step => step.key === key);
    const step = steps[index];
    if (step.id) {
      const newStepsToDelete = [ ...stepsToDelete ];
      newStepsToDelete.push(step.id);
      setStepsToDelete(newStepsToDelete);
    }

    setSteps(
      steps.slice(0, index).concat(steps.slice(index + 1))
    );
  };

  const errors = props.errors;
  return (
    <div className="row mb-4">
      <label htmlFor="newMethodStep" className="col-md-3 col-form-label text-md-end">Method steps</label>
      <div className="col-md-9">
        <div className="row">
          {stepsToDelete.map(stepId => (
            <input key={stepId}
                   type="hidden"
                   name="stepsToDelete"
                   value={stepId}
            />
          ))}
        </div>

        {steps.map(step => (
          <div key={step.key} className="row">
            <MethodStepFields
              step={step}
              onUpdate={(step) => updateStep(step)}
              onDelete={() => deleteStep(step.key)}
            />
          </div>
        ))}

        <div className="row">
          <NewMethodStepFields value={newStep} onChange={setNewStep} onSubmit={addStep} />
        </div>

        <div className="row">
          {errors && errors.map(error => {
            return (
              <div key={error} className="invalid-feedback d-block">
                {error}
              </div>
            )
          })}
        </div>
      </div>
    </div>
  );
};

type MethodStepFieldsProps = {
  step: StepComponentData;
  onDelete: () => void;
  onUpdate: (step: StepComponentData) => void;
};

const MethodStepFields = (props: MethodStepFieldsProps) => {
  const { step, onDelete, onUpdate } = props;
  const { id } = step;

  return (
    <div className="input-group">
      <input
        type="hidden"
        name="methodStepId"
        value={id}
      />

      <textarea
        name="methodStepText"
        className="form-control"
        placeholder="Method step"
        value={step.text}
        onChange={(event) => onUpdate({ ...step, text: event.target.value })}
        style={{height: "100px"}}
      />

      <div className="input-group-text">
        <button className="btn btn-sm btn-outline-danger" type="button" onClick={onDelete}>
          <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor"
               className="bi bi-dash-lg" viewBox="0 0 16 16">
            <path fillRule="evenodd" d="M2 8a.5.5 0 0 1 .5-.5h11a.5.5 0 0 1 0 1h-11A.5.5 0 0 1 2 8"/>
          </svg>
        </button>
      </div>
    </div>
  );
};

const NewMethodStepFields = (
  props: {
    value: string,
    onChange: (value: string) => void,
    onSubmit: () => void,
  }
) => {
  const {value, onChange, onSubmit} = props;
  return (
    <div className="input-group">
      <textarea
        name="newMethodStep" id="newMethodStep" className="form-control"
        value={value}
        placeholder="Instructions step-by-step to cook your recipe"
        style={{height: "100px"}}
        onChange={(e) => onChange(e.currentTarget.value)}
        onKeyDown={(e) => {
          if (e.key === 'Enter') {
            e.preventDefault();
            onSubmit();
          }
        }}
      />
      <div className="input-group-text">
        <button className="btn btn-sm btn-success" type="button" onClick={onSubmit}>
          <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor"
               className="bi bi-plus-lg" viewBox="0 0 16 16">
            <path fillRule="evenodd"
                  d="M8 2a.5.5 0 0 1 .5.5v5h5a.5.5 0 0 1 0 1h-5v5a.5.5 0 0 1-1 0v-5h-5a.5.5 0 0 1 0-1h5v-5A.5.5 0 0 1 8 2"/>
          </svg>
        </button>
      </div>
    </div>
  );
};

export {MethodStepsInput};