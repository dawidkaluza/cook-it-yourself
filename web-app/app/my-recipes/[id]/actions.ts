"use server";

import {RecipeDetails} from "@/app/my-recipes/_dtos/recipe";
import { fetchFromServer} from "@/app/_api/fetch";

export async function reviewRecipe(id: number) : Promise<RecipeDetails> {
  return await fetchFromServer<RecipeDetails>({
    endpoint: `/kitchen/recipe/${id}`
  }) as RecipeDetails;
}