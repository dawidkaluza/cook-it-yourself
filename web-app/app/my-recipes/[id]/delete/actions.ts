"use server";

import {ApiError, fetchFromServer} from "@/app/_api/fetch";
import {redirect} from "next/navigation";

export async function deleteRecipe(prevState: any, formData: FormData): Promise<string> {
  const id = formData.get("id");

  try {
    await fetchFromServer({
      endpoint: `/kitchen/recipe/${id}`,
      method: "DELETE"
    });
  } catch (error) {
    if (error instanceof ApiError) {
      const response = error.response;
      switch (response.status) {
        case 404: {
          return "The recipe is already deleted.";
        }

        case 422: {
          return "Couldn't process the request.";
        }
      }
    }

    throw error;
  }

  redirect("/my-recipes");
}