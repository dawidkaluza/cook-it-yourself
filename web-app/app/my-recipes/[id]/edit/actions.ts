"use server";

import {FieldError} from "@/app/my-recipes/_dtos/errors";
import {ApiError, fetchFromServer} from "@/app/_api/fetch";
import {PortionSize, RecipeDetails, UpdateRecipeRequest} from "@/app/my-recipes/_dtos/recipe";
import {redirect} from "next/navigation";

async function editRecipe(prevState: any, formData: FormData): Promise<FieldError[]> {
  const id = formData.get("id") as string;

  const mapToPortionSize = (formData: FormData): PortionSize => {
    const portionSizeInput = formData?.get("portionSize")?.toString();
    if (!portionSizeInput) {
      return {
        value: "",
        measure: "",
      };
    }

    const numberIndex = portionSizeInput.search(/\d/g);
    if (numberIndex === -1) {
      return {
        value: "",
        measure: "",
      };
    }

    const nonNumberIndex = portionSizeInput.search(/\D/g);
    if (nonNumberIndex === -1 || nonNumberIndex < numberIndex) {
      return {
        value: portionSizeInput.slice(numberIndex),
        measure: "",
      };
    }

    return {
      value: portionSizeInput.slice(numberIndex, nonNumberIndex).trim(),
      measure: portionSizeInput.slice(nonNumberIndex).trim()
    };
  }

  const requestData: UpdateRecipeRequest = {
    basicInformation: {
      name: formData.get("name") as string,
      description: formData.get("description") as string,
      cookingTime: Number(formData.get("cookingTime")),
      portionSize: mapToPortionSize(formData),
    },
    ingredients: {
      ingredientsToAdd: [],
      ingredientsToUpdate: [],
      ingredientsToDelete: [],
    },
    steps: {
      stepsToAdd: [],
      stepsToUpdate: [],
      stepsToDelete: [],
    }
  };

  try {
    const updatedRecipe = await fetchFromServer<RecipeDetails>({
      endpoint: `/kitchen/recipe/${id}`,
      method: "PUT",
      body: requestData
    });

    if (updatedRecipe) {
      redirect(`/my-recipes/${updatedRecipe.id}/`);
    }
  } catch (error) {
    if (error instanceof ApiError) {
      const response = error.response;

      // TODO cover other possible client errors
      if (response.status === 422) {
        const body = await response.json();
        const fields : FieldError[] = body.fields;
        return fields.map((field) => {
          const name = field.name;
          const message = field.message;
          const indexOfDot = name.indexOf(".");
          if (indexOfDot === -1) {
            return {
              name,
              message
            }
          } else {
            const mainName = name.slice(0, indexOfDot);
            return {
              name: mainName,
              message
            };
          }
        });
      }

      return [];
    }

    throw error;
  }

  throw new Error("Couldn't process the request");
}

export { editRecipe };