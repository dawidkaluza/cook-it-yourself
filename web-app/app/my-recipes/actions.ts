"use server";

import {fetchFromServer} from "@/app/_api/fetch";
import {Page} from "@/app/my-recipes/_dtos/page";
import {Recipe} from "@/app/my-recipes/_dtos/recipe";

export async function getMyRecipes(){
  return await fetchFromServer({ endpoint: "/kitchen/recipe" }) as Promise<Page<Recipe>>;
}