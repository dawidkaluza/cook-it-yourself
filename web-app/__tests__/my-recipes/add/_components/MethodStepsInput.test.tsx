import {MethodStepsInput} from "@/app/my-recipes/add/_components/MethodStepsInput";
import {render, screen, within} from "@testing-library/react";
import {FieldError} from "@/app/my-recipes/add/actions";
import userEvent from "@testing-library/user-event";

describe("MethodStepsInput component", () => {
  test("render without validation errors", () => {
    // Given
    // When
    const component = render(<MethodStepsInput />);

    // Then
    const inputElement = screen.getByLabelText("Method steps");
    expect(inputElement).not.toBeNull();

    expect(component.container).toMatchSnapshot();
  });

  test("render with validation errors", () => {
    // Given
    const fieldErrors: FieldError[] = [
      {
        name: "methodStep",
        message: "Some text error.",
      },
      {
        name: "methodStep",
        message: "Some other text error.",
      }
    ];

    // When
    const component = render(<MethodStepsInput fieldErrors={fieldErrors} />);

    // Then
    const inputElement = screen.getByLabelText("Method steps");
    expect(inputElement).not.toBeNull();

    for (const fieldError of fieldErrors) {
      const errorElement = screen.getByText(fieldError.message);
      expect(errorElement).not.toBeNull();
    }

    expect(component.container).toMatchSnapshot();
  });

  test("render and add new step", async () => {
    // Given
    const expectedInputs = [
      "Add some sugar",
      "Mix everything",
      "Pour water into the pot"
    ];

    const user = userEvent.setup();
    const component = render(<MethodStepsInput />);

    // When
    const newStepInputEl = screen.getByLabelText("Method steps");
    const addBtnEl = screen.getByRole("button");

    await user.clear(newStepInputEl);
    await user.click(addBtnEl);

    await user.type(newStepInputEl, "Add some sugar");
    await user.click(addBtnEl);

    await user.type(newStepInputEl, "Mix everything");
    await user.click(addBtnEl);

    await user.type(newStepInputEl, "Pour water into the pot");
    await user.click(newStepInputEl);
    await user.keyboard("{Enter}");

    // Then
    const allInputEls = screen.getAllByRole("textbox") as HTMLTextAreaElement[];
    for (const expectedInput of expectedInputs) {
      expect(allInputEls.filter(
        el => el.value === expectedInput
      )).toHaveLength(1);
    }

    expect(component.container).toMatchSnapshot();
  });

  test("render, add and modify added steps", async () => {
    // Given
    const expectedInputs = [
      "Add some salt",
      "Mix everything",
      "Add some oil"
    ];

    const user = userEvent.setup();
    const component = render(<MethodStepsInput />);

    const newStepInputEl = screen.getByLabelText("Method steps");
    const addBtnEl = screen.getByRole("button");

    await user.type(newStepInputEl, "Add some sugar");
    await user.click(addBtnEl);

    await user.type(newStepInputEl, "Mix everything");
    await user.click(addBtnEl);

    await user.type(newStepInputEl, "Pour water into the pot");
    await user.click(addBtnEl);

    // When
    const addSugarStepEl = await screen.getByDisplayValue("Add some sugar");
    await user.clear(addSugarStepEl);
    await user.type(addSugarStepEl, "Add some salt");

    const pourWaterStepEl = await screen.getByDisplayValue("Pour water into the pot");
    await user.clear(pourWaterStepEl);
    await user.type(pourWaterStepEl, "Add some oil");

    // Then
    const allInputEls = screen.getAllByRole("textbox") as HTMLTextAreaElement[];
    for (const expectedInput of expectedInputs) {
      expect(allInputEls.filter(
        el => el.value === expectedInput
      )).toHaveLength(1);
    }

    expect(component.container).toMatchSnapshot();
  });

  test("render, add and remove steps", async () => {
    // Given
    const expectedInputs = [
      "Add some sugar",
      "Pour water into the pot"
    ];

    const expectedRemovedInputs = [
      "Mix everything",
    ];

    const user = userEvent.setup();
    const component = render(<MethodStepsInput />);

    const newStepInputEl = screen.getByLabelText("Method steps");
    const addBtnEl = screen.getByRole("button");

    await user.type(newStepInputEl, "Add some sugar");
    await user.click(addBtnEl);

    await user.type(newStepInputEl, "Mix everything");
    await user.click(addBtnEl);

    await user.type(newStepInputEl, "Pour water into the pot");
    await user.click(addBtnEl);

    // When
    const mixDeleteBtnEl = within(
      screen.getByDisplayValue("Mix everything").parentElement as HTMLElement
    ).getByRole("button");
    await user.click(mixDeleteBtnEl);

    // Then
    const allInputEls = screen.getAllByRole("textbox") as HTMLTextAreaElement[];

    for (const expectedInput of expectedInputs) {
      expect(allInputEls.filter(
        el => el.value === expectedInput
      )).toHaveLength(1);
    }

    for (const expectedRemovedInput of expectedRemovedInputs) {
      expect(allInputEls.filter(
        el => el.value === expectedRemovedInput
      )).toHaveLength(0);
    }

    expect(component.container).toMatchSnapshot();
  });
});