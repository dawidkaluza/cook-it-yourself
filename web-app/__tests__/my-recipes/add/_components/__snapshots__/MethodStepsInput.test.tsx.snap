// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MethodStepsInput component > render and add new step 1`] = `
<div>
  <div
    class="row mb-4"
  >
    <label
      class="col-md-3 col-form-label text-md-end"
      for="newMethodStep"
    >
      Method steps
    </label>
    <div
      class="col-md-9"
    >
      <div
        class="row"
      >
        <div
          class="input-group"
        >
          <textarea
            class="form-control"
            name="methodSteps"
            placeholder="Method step"
            style="height: 100px;"
          >
            Add some sugar
          </textarea>
          <div
            class="input-group-text"
          >
            <button
              class="btn btn-sm btn-outline-danger"
              type="button"
            >
              <svg
                class="bi bi-dash-lg"
                fill="currentColor"
                height="16"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M2 8a.5.5 0 0 1 .5-.5h11a.5.5 0 0 1 0 1h-11A.5.5 0 0 1 2 8"
                  fill-rule="evenodd"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="input-group"
        >
          <textarea
            class="form-control"
            name="methodSteps"
            placeholder="Method step"
            style="height: 100px;"
          >
            Mix everything
          </textarea>
          <div
            class="input-group-text"
          >
            <button
              class="btn btn-sm btn-outline-danger"
              type="button"
            >
              <svg
                class="bi bi-dash-lg"
                fill="currentColor"
                height="16"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M2 8a.5.5 0 0 1 .5-.5h11a.5.5 0 0 1 0 1h-11A.5.5 0 0 1 2 8"
                  fill-rule="evenodd"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="input-group"
        >
          <textarea
            class="form-control"
            name="methodSteps"
            placeholder="Method step"
            style="height: 100px;"
          >
            Pour water into the pot
          </textarea>
          <div
            class="input-group-text"
          >
            <button
              class="btn btn-sm btn-outline-danger"
              type="button"
            >
              <svg
                class="bi bi-dash-lg"
                fill="currentColor"
                height="16"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M2 8a.5.5 0 0 1 .5-.5h11a.5.5 0 0 1 0 1h-11A.5.5 0 0 1 2 8"
                  fill-rule="evenodd"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="input-group"
        >
          <textarea
            class="form-control"
            id="newMethodStep"
            name="newMethodStep"
            placeholder="Instructions step-by-step to cook your recipe"
            style="height: 100px;"
          />
          <div
            class="input-group-text"
          >
            <button
              class="btn btn-sm btn-success"
              type="button"
            >
              <svg
                class="bi bi-plus-lg"
                fill="currentColor"
                height="16"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8 2a.5.5 0 0 1 .5.5v5h5a.5.5 0 0 1 0 1h-5v5a.5.5 0 0 1-1 0v-5h-5a.5.5 0 0 1 0-1h5v-5A.5.5 0 0 1 8 2"
                  fill-rule="evenodd"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <div
        class="row"
      />
    </div>
  </div>
</div>
`;

exports[`MethodStepsInput component > render with validation errors 1`] = `
<div>
  <div
    class="row mb-4"
  >
    <label
      class="col-md-3 col-form-label text-md-end"
      for="newMethodStep"
    >
      Method steps
    </label>
    <div
      class="col-md-9"
    >
      <div
        class="row"
      >
        <div
          class="input-group"
        >
          <textarea
            class="form-control"
            id="newMethodStep"
            name="newMethodStep"
            placeholder="Instructions step-by-step to cook your recipe"
            style="height: 100px;"
          />
          <div
            class="input-group-text"
          >
            <button
              class="btn btn-sm btn-success"
              type="button"
            >
              <svg
                class="bi bi-plus-lg"
                fill="currentColor"
                height="16"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8 2a.5.5 0 0 1 .5.5v5h5a.5.5 0 0 1 0 1h-5v5a.5.5 0 0 1-1 0v-5h-5a.5.5 0 0 1 0-1h5v-5A.5.5 0 0 1 8 2"
                  fill-rule="evenodd"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="invalid-feedback d-block"
        >
          Some text error.
        </div>
        <div
          class="invalid-feedback d-block"
        >
          Some other text error.
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`MethodStepsInput component > render without validation errors 1`] = `
<div>
  <div
    class="row mb-4"
  >
    <label
      class="col-md-3 col-form-label text-md-end"
      for="newMethodStep"
    >
      Method steps
    </label>
    <div
      class="col-md-9"
    >
      <div
        class="row"
      >
        <div
          class="input-group"
        >
          <textarea
            class="form-control"
            id="newMethodStep"
            name="newMethodStep"
            placeholder="Instructions step-by-step to cook your recipe"
            style="height: 100px;"
          />
          <div
            class="input-group-text"
          >
            <button
              class="btn btn-sm btn-success"
              type="button"
            >
              <svg
                class="bi bi-plus-lg"
                fill="currentColor"
                height="16"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8 2a.5.5 0 0 1 .5.5v5h5a.5.5 0 0 1 0 1h-5v5a.5.5 0 0 1-1 0v-5h-5a.5.5 0 0 1 0-1h5v-5A.5.5 0 0 1 8 2"
                  fill-rule="evenodd"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <div
        class="row"
      />
    </div>
  </div>
</div>
`;

exports[`MethodStepsInput component > render, add and modify added steps 1`] = `
<div>
  <div
    class="row mb-4"
  >
    <label
      class="col-md-3 col-form-label text-md-end"
      for="newMethodStep"
    >
      Method steps
    </label>
    <div
      class="col-md-9"
    >
      <div
        class="row"
      >
        <div
          class="input-group"
        >
          <textarea
            class="form-control"
            name="methodSteps"
            placeholder="Method step"
            style="height: 100px;"
          >
            Add some salt
          </textarea>
          <div
            class="input-group-text"
          >
            <button
              class="btn btn-sm btn-outline-danger"
              type="button"
            >
              <svg
                class="bi bi-dash-lg"
                fill="currentColor"
                height="16"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M2 8a.5.5 0 0 1 .5-.5h11a.5.5 0 0 1 0 1h-11A.5.5 0 0 1 2 8"
                  fill-rule="evenodd"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="input-group"
        >
          <textarea
            class="form-control"
            name="methodSteps"
            placeholder="Method step"
            style="height: 100px;"
          >
            Mix everything
          </textarea>
          <div
            class="input-group-text"
          >
            <button
              class="btn btn-sm btn-outline-danger"
              type="button"
            >
              <svg
                class="bi bi-dash-lg"
                fill="currentColor"
                height="16"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M2 8a.5.5 0 0 1 .5-.5h11a.5.5 0 0 1 0 1h-11A.5.5 0 0 1 2 8"
                  fill-rule="evenodd"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="input-group"
        >
          <textarea
            class="form-control"
            name="methodSteps"
            placeholder="Method step"
            style="height: 100px;"
          >
            Add some oil
          </textarea>
          <div
            class="input-group-text"
          >
            <button
              class="btn btn-sm btn-outline-danger"
              type="button"
            >
              <svg
                class="bi bi-dash-lg"
                fill="currentColor"
                height="16"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M2 8a.5.5 0 0 1 .5-.5h11a.5.5 0 0 1 0 1h-11A.5.5 0 0 1 2 8"
                  fill-rule="evenodd"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="input-group"
        >
          <textarea
            class="form-control"
            id="newMethodStep"
            name="newMethodStep"
            placeholder="Instructions step-by-step to cook your recipe"
            style="height: 100px;"
          />
          <div
            class="input-group-text"
          >
            <button
              class="btn btn-sm btn-success"
              type="button"
            >
              <svg
                class="bi bi-plus-lg"
                fill="currentColor"
                height="16"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8 2a.5.5 0 0 1 .5.5v5h5a.5.5 0 0 1 0 1h-5v5a.5.5 0 0 1-1 0v-5h-5a.5.5 0 0 1 0-1h5v-5A.5.5 0 0 1 8 2"
                  fill-rule="evenodd"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <div
        class="row"
      />
    </div>
  </div>
</div>
`;

exports[`MethodStepsInput component > render, add and remove steps 1`] = `
<div>
  <div
    class="row mb-4"
  >
    <label
      class="col-md-3 col-form-label text-md-end"
      for="newMethodStep"
    >
      Method steps
    </label>
    <div
      class="col-md-9"
    >
      <div
        class="row"
      >
        <div
          class="input-group"
        >
          <textarea
            class="form-control"
            name="methodSteps"
            placeholder="Method step"
            style="height: 100px;"
          >
            Add some sugar
          </textarea>
          <div
            class="input-group-text"
          >
            <button
              class="btn btn-sm btn-outline-danger"
              type="button"
            >
              <svg
                class="bi bi-dash-lg"
                fill="currentColor"
                height="16"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M2 8a.5.5 0 0 1 .5-.5h11a.5.5 0 0 1 0 1h-11A.5.5 0 0 1 2 8"
                  fill-rule="evenodd"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="input-group"
        >
          <textarea
            class="form-control"
            name="methodSteps"
            placeholder="Method step"
            style="height: 100px;"
          >
            Pour water into the pot
          </textarea>
          <div
            class="input-group-text"
          >
            <button
              class="btn btn-sm btn-outline-danger"
              type="button"
            >
              <svg
                class="bi bi-dash-lg"
                fill="currentColor"
                height="16"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M2 8a.5.5 0 0 1 .5-.5h11a.5.5 0 0 1 0 1h-11A.5.5 0 0 1 2 8"
                  fill-rule="evenodd"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="input-group"
        >
          <textarea
            class="form-control"
            id="newMethodStep"
            name="newMethodStep"
            placeholder="Instructions step-by-step to cook your recipe"
            style="height: 100px;"
          />
          <div
            class="input-group-text"
          >
            <button
              class="btn btn-sm btn-success"
              type="button"
            >
              <svg
                class="bi bi-plus-lg"
                fill="currentColor"
                height="16"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8 2a.5.5 0 0 1 .5.5v5h5a.5.5 0 0 1 0 1h-5v5a.5.5 0 0 1-1 0v-5h-5a.5.5 0 0 1 0-1h5v-5A.5.5 0 0 1 8 2"
                  fill-rule="evenodd"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <div
        class="row"
      />
    </div>
  </div>
</div>
`;
